### PR TITLE
Fix type for memcache get

### PR DESCRIPTION
**Pull request by jarstelfox**
See docs:
* https://www.php.net/manual/en/memcached.get.php
* https://www.php.net/manual/en/memcached.getresultcode.php

### CODE REVIEW

File: library/iFixit/Matryoshka/Memcached.php [hunk 1]
```diff
@@ -60,10 +60,12 @@ public function decrement($key, $amount = 1, $expiration = 0) {
        $expiration);
    }
 
+   /** @return mixed|null */
```
Comment (andyg0808) [inline library/iFixit/Matryoshka/Memcached.php]: Can you use
```suggestion
   /** @return mixed|self::MISS */
```

**Comment by andyg0808** [comment]
CR :+1:
deploy_block :+1:
on comment

**Comment by djmetzle** [comment]
Nice job running this down! CR :clamp:

File: library/iFixit/Matryoshka/Memcached.php (hunk 1 above)
Comment (jarstelfox) [reply library/iFixit/Matryoshka/Memcached.php]: Cool! I didn't know that was possible: https://psalm.dev/r/a3bb06fde6

**Comment by andyg0808** [comment]
CR :+1:

**Comment by danielbeardsley** [comment]
@jarstelfox and I talked. I think the best option is to throw an exception. If we pretend errors are misses, and we had another network event as we did last night, it would rather instantly overload our DB.

Let's throw an exception cause these cases are seemingly rare. If we want to do something like: Pretend the first error of this type is a miss but throw an exception on the second instance during a request, we can do that as a Matryoshka layer in our main application.

dev_block 👍  on throwing an exception in all exceptional cases (everything besides found / not found).

**Comment by jarstelfox** [comment]
un_dev_block 🌵

File: library/iFixit/Matryoshka/Memcached.php [hunk 2]
```diff
@@ -60,16 +89,38 @@ public function decrement($key, $amount = 1, $expiration = 0) {
...
+      $result = $this->memcached->getResultCode();
 
-      if ($this->memcached->getResultCode() === \Memcached::RES_NOTFOUND) {
+      if ($this->isMiss($result)) {
          return self::MISS;
       }
 
+      $error = $this->getError($result);
+      if ($error) {
+         throw $error;
```
Comment (djmetzle) [inline library/iFixit/Matryoshka/Memcached.php:107]: This is kind of surprising behavior change in a public library.

But i think it's a valid one.

File: library/iFixit/Matryoshka/Memcached.php [hunk 3]
```diff
@@ -60,16 +89,38 @@ public function decrement($key, $amount = 1, $expiration = 0) {
...
       return $value;
    }
 
+   private function isMiss(int $resultCode) {
+      return $resultCode === \Memcached::RES_NOTFOUND;
+   }
+
+   private function getError(int $resultCode): ?\MemcachedException {
+      $hasKey = array_key_exists($resultCode, self::ERRORS);
+      $errorMessage = $hasKey ? self::ERRORS[$resultCode] : null;
```
Comment (djmetzle) [inline library/iFixit/Matryoshka/Memcached.php:120]: Nah, lets have a sane default message here. Empty exception strings are _weird_...

```suggestion
      $errorMessage = $hasKey
       ? self::ERRORS[$resultCode]
       : "An unknown error occured: {$resultCode}";
```

What do you think of that? I'm not tied to that exact message, but it should be something helpful...

Comment (djmetzle) [reply library/iFixit/Matryoshka/Memcached.php:120]: We run mcrouter in prod, which might have it's own special error numbers...

Comment (jarstelfox) [reply library/iFixit/Matryoshka/Memcached.php:120]: Oh right now I also pass success into this function - it's just an error code. I have covered the width of errors in the array. I don't mind making found/not found early exists, then a default error message. Opinions?

Comment (djmetzle) [reply library/iFixit/Matryoshka/Memcached.php:120]: Huh? I'm just talking about the default exception message if we don't know about a specific error number.

Comment (danielbeardsley) [reply library/iFixit/Matryoshka/Memcached.php:120]: If we are going to treat an unknown result code as an error, we'll have to early exit if `resultCode == success`.

**Comment by djmetzle** [comment]
CR :gear:

**Comment by djmetzle** [comment]
CR :gear:

File: library/iFixit/Matryoshka/Memcached.php [hunk 4]
```diff
@@ -60,14 +60,34 @@ public function decrement($key, $amount = 1, $expiration = 0) {
...
+    * @template T
+    * @return T|self::MISS
+    */
    public function get($key) {
+      /** @var T|false */
       $value = $this->memcached->get($key);
+      $result = $this->memcached->getResultCode();
+
+      if ($this->isSuccess($result)) {
+         /** @var T */
```
Comment (danielbeardsley) [inline library/iFixit/Matryoshka/Memcached.php:73]: If you *store* false... does memcache later return null with a `MISS` result code? or does it still return `false` with a `SUCCESS` code?

I feel like we should have a test around this.

**Comment by danielbeardsley** [comment]
CR 👍 

But deploy_block 👍  on my comment. 

this is looking great though!

File: library/iFixit/Matryoshka/Memcached.php (hunk 4 above)
Comment (jarstelfox) [reply library/iFixit/Matryoshka/Memcached.php:73]: RIght now `false` will be returned if that was cached. A miss is only considering the result code, not the stored value. I.E.yes, it is possible to store false in memcache, ask for it, and have it be a successful call

**Comment by danielbeardsley** [comment]
CR 👍 

un_deploy_block 👍

**Comment by danielbeardsley** [comment]
When we utilize this, we'll definitely want a error-swallowing backend. Right now, we don't know how frequently an error is taken as a miss and the app carries on without further issues. I wouldn't want all those to turn into 500s.

**Comment by djmetzle** [comment]
CR :woman_technologist:

File: library/iFixit/Matryoshka/Memcached.php (hunk 4 above)
Comment (jarstelfox) [reply library/iFixit/Matryoshka/Memcached.php:73]: >I feel like we should have a test around this.

We already do 🎉

**Pull request merged**